From cc430129567a3f0d2f2acca1596e2f1a81c6611c Mon Sep 17 00:00:00 2001
From: Tai Huynh <tai.huynh.cp@renesas.com>
Date: Fri, 25 Oct 2024 14:53:42 +0700
Subject: [PATCH 2/3] net: ethernet: renesas: ravb: add code to access over 32
 bits memory region for RZ/V2M

RZ/V2M only supports over 32 bits memory region in Linux.
To support ethernet can work with this region, extended memory
access is applied although RZ/V2M Ethernet controller can not access to
64 bits memory.
It will switch bank 0 (0x0 - 0xFFFF_FFFF) address to
bank 1 address (0x1_0000_0000 - 0x1_FFFF_FFFF) by SYS that was set
before.

Signed-off-by: Tai Huynh <tai.huynh.cp@renesas.com>
---
 drivers/net/ethernet/renesas/ravb.h      |  2 +
 drivers/net/ethernet/renesas/ravb_main.c | 90 ++++++++++++++++++++----
 2 files changed, 80 insertions(+), 12 deletions(-)

diff --git a/drivers/net/ethernet/renesas/ravb.h b/drivers/net/ethernet/renesas/ravb.h
index 1abb541a6829..88d47cfd7a37 100644
--- a/drivers/net/ethernet/renesas/ravb.h
+++ b/drivers/net/ethernet/renesas/ravb.h
@@ -1057,6 +1057,8 @@ struct ravb_hw_info {
 	u32 rx_max_desc_use;
 	u32 rx_desc_size;
 	unsigned aligned_tx: 1;
+	unsigned use_64_bit_dma: 1;     /* Use 64-bit for DMA descriptor */
+	u64 extended_addr;              /* Use 64-bit for DMA descriptor */
 
 	/* hardware features */
 	unsigned internal_delay:1;	/* AVB-DMAC has internal delays */
diff --git a/drivers/net/ethernet/renesas/ravb_main.c b/drivers/net/ethernet/renesas/ravb_main.c
index e01f933ff74e..de51534bc965 100644
--- a/drivers/net/ethernet/renesas/ravb_main.c
+++ b/drivers/net/ethernet/renesas/ravb_main.c
@@ -210,6 +210,7 @@ ravb_rx_get_desc(struct ravb_private *priv, unsigned int q,
 static int ravb_tx_free(struct net_device *ndev, int q, bool free_txed_only)
 {
 	struct ravb_private *priv = netdev_priv(ndev);
+	const struct ravb_hw_info *info = priv->info;
 	struct net_device_stats *stats = &priv->stats[q];
 	unsigned int num_tx_desc = priv->num_tx_desc;
 	struct ravb_tx_desc *desc;
@@ -231,6 +232,10 @@ static int ravb_tx_free(struct net_device *ndev, int q, bool free_txed_only)
 		size = le16_to_cpu(desc->ds_tagl) & TX_DS;
 		/* Free the original skb. */
 		if (priv->tx_skb[q][entry / num_tx_desc]) {
+		if (info->use_64_bit_dma) /* Try 64-bit mask if hardware is capable of it */
+		        dma_unmap_single(ndev->dev.parent, info->extended_addr | le32_to_cpu(desc->dptr),
+		                          size, DMA_TO_DEVICE);
+		else
 			dma_unmap_single(ndev->dev.parent, le32_to_cpu(desc->dptr),
 					 size, DMA_TO_DEVICE);
 			/* Last packet descriptor? */
@@ -253,6 +258,7 @@ static int ravb_tx_free(struct net_device *ndev, int q, bool free_txed_only)
 static void ravb_rx_ring_free(struct net_device *ndev, int q)
 {
 	struct ravb_private *priv = netdev_priv(ndev);
+	const struct ravb_hw_info *info = priv->info;
 	unsigned int ring_size;
 	unsigned int i;
 
@@ -263,11 +269,19 @@ static void ravb_rx_ring_free(struct net_device *ndev, int q)
 		struct ravb_rx_desc *desc = ravb_rx_get_desc(priv, q, i);
 
 		if (!dma_mapping_error(ndev->dev.parent,
-				       le32_to_cpu(desc->dptr)))
-			dma_unmap_single(ndev->dev.parent,
-					 le32_to_cpu(desc->dptr),
-					 priv->info->rx_max_frame_size,
-					 DMA_FROM_DEVICE);
+				       le32_to_cpu(desc->dptr))) {
+			if (info->use_64_bit_dma)
+				/* Try 64-bit mask if hardware is capable of it */
+				dma_unmap_single(ndev->dev.parent,
+						 info->extended_addr | le32_to_cpu(desc->dptr),
+						 priv->info->rx_max_frame_size,
+						 DMA_FROM_DEVICE);
+			else
+				dma_unmap_single(ndev->dev.parent,
+						 le32_to_cpu(desc->dptr),
+						 priv->info->rx_max_frame_size,
+						 DMA_FROM_DEVICE);
+		}
 	}
 	ring_size = priv->info->rx_desc_size * (priv->num_rx_ring[q] + 1);
 	dma_free_coherent(ndev->dev.parent, ring_size, priv->rx_ring[q].raw,
@@ -879,6 +893,7 @@ static int ravb_rx_gbeth(struct net_device *ndev, int budget, int q)
 static int ravb_rx_rcar(struct net_device *ndev, int budget, int q)
 {
 	struct ravb_private *priv = netdev_priv(ndev);
+	const struct ravb_hw_info *info = priv->info;
 	int entry;
 	struct net_device_stats *stats = &priv->stats[q];
 	struct ravb_ex_rx_desc *desc;
@@ -924,9 +939,14 @@ static int ravb_rx_rcar(struct net_device *ndev, int budget, int q)
 
 			skb = priv->rx_skb[q][entry];
 			priv->rx_skb[q][entry] = NULL;
-			dma_unmap_single(ndev->dev.parent, le32_to_cpu(desc->dptr),
-					 priv->info->rx_max_frame_size,
-					 DMA_FROM_DEVICE);
+			if (info->use_64_bit_dma) /* Try 64-bit mask if hardware is capable of it */
+			        dma_unmap_single(ndev->dev.parent, info->extended_addr | le32_to_cpu(desc->dptr),
+			                         priv->info->rx_max_frame_size,
+			                         DMA_FROM_DEVICE);
+			else
+			        dma_unmap_single(ndev->dev.parent, le32_to_cpu(desc->dptr),
+			                         priv->info->rx_max_frame_size,
+			                         DMA_FROM_DEVICE);
 			get_ts &= (q == RAVB_NC) ?
 					RAVB_RXTSTAMP_TYPE_V2_L2_EVENT :
 					~RAVB_RXTSTAMP_TYPE_V2_L2_EVENT;
@@ -1923,7 +1943,7 @@ static netdev_tx_t ravb_start_xmit(struct sk_buff *skb, struct net_device *ndev)
 	struct ravb_tstamp_skb *ts_skb;
 	struct ravb_tx_desc *desc;
 	unsigned long flags;
-	u32 dma_addr;
+	dma_addr_t dma_addr;
 	void *buffer;
 	u32 entry;
 	u32 len;
@@ -1977,6 +1997,9 @@ static netdev_tx_t ravb_start_xmit(struct sk_buff *skb, struct net_device *ndev)
 
 		desc = &priv->tx_ring[q][entry];
 		desc->ds_tagl = cpu_to_le16(len);
+		if (info->use_64_bit_dma) /* Try 64-bit mask if hardware is capable of it */
+		desc->dptr = cpu_to_le32((u32)dma_addr);
+		else
 		desc->dptr = cpu_to_le32(dma_addr);
 
 		buffer = skb->data + len;
@@ -2045,8 +2068,12 @@ static netdev_tx_t ravb_start_xmit(struct sk_buff *skb, struct net_device *ndev)
 	return NETDEV_TX_OK;
 
 unmap:
-	dma_unmap_single(ndev->dev.parent, le32_to_cpu(desc->dptr),
-			 le16_to_cpu(desc->ds_tagl), DMA_TO_DEVICE);
+	if (info->use_64_bit_dma) /* Try 64-bit mask if hardware is capable of it */
+	        dma_unmap_single(ndev->dev.parent, info->extended_addr | le32_to_cpu(desc->dptr),
+			le16_to_cpu(desc->ds_tagl), DMA_TO_DEVICE);
+	else
+	        dma_unmap_single(ndev->dev.parent, le32_to_cpu(desc->dptr),
+			le16_to_cpu(desc->ds_tagl), DMA_TO_DEVICE);
 drop:
 	dev_kfree_skb_any(skb);
 	priv->tx_skb[q][entry / num_tx_desc] = NULL;
@@ -2511,6 +2538,30 @@ static const struct ravb_hw_info ravb_rzv2m_hw_info = {
 	.gptp_ref_clk = 1,
 	.nc_queues = 1,
 	.magic_pkt = 1,
+	.use_64_bit_dma = 1,
+};
+
+static const struct ravb_hw_info ravb_rzv2ma_hw_info = {
+        .receive = ravb_rx_rcar,
+        .set_rate = ravb_set_rate_rcar,
+        .set_feature = ravb_set_features_rcar,
+        .dmac_init = ravb_dmac_init_rcar,
+        .emac_init = ravb_emac_init_rcar,
+        .gstrings_stats = ravb_gstrings_stats,
+        .gstrings_size = sizeof(ravb_gstrings_stats),
+        .net_hw_features = NETIF_F_RXCSUM,
+        .net_features = NETIF_F_RXCSUM,
+        .stats_len = ARRAY_SIZE(ravb_gstrings_stats),
+        .tccr_mask = TCCR_TSRQ0 | TCCR_TSRQ1 | TCCR_TSRQ2 | TCCR_TSRQ3,
+        .rx_max_frame_size = SZ_2K,
+        .rx_max_desc_use = SZ_2K - ETH_FCS_LEN + sizeof(__sum16),
+        .rx_desc_size = sizeof(struct ravb_ex_rx_desc),
+        .multi_irqs = 1,
+        .err_mgmt_irqs = 1,
+        .gptp = 1,
+        .gptp_ref_clk = 1,
+        .nc_queues = 1,
+        .magic_pkt = 1,
 };
 
 static const struct ravb_hw_info gbeth_hw_info = {
@@ -2546,7 +2597,7 @@ static const struct of_device_id ravb_match_table[] = {
 	{ .compatible = "renesas,etheravb-r8a7795", .data = &ravb_gen3_hw_info },
 	{ .compatible = "renesas,etheravb-rcar-gen3", .data = &ravb_gen3_hw_info },
 	{ .compatible = "renesas,etheravb-rzv2m", .data = &ravb_rzv2m_hw_info },
-	{ .compatible = "renesas,etheravb-rzv2ma", .data = &ravb_rzv2m_hw_info },
+	{ .compatible = "renesas,etheravb-rzv2ma", .data = &ravb_rzv2ma_hw_info },
 	{ .compatible = "renesas,rzg2l-gbeth", .data = &gbeth_hw_info },
 	{ }
 };
@@ -2674,6 +2725,21 @@ static int ravb_probe(struct platform_device *pdev)
 
 	info = of_device_get_match_data(&pdev->dev);
 
+	/* Try 64-bit mask if hardware is capable of it */
+	if (info->use_64_bit_dma) {
+	        error = dma_set_mask_and_coherent(&pdev->dev, DMA_BIT_MASK(64));
+	        if (error)
+	                goto out_free_netdev;
+	
+	        error = of_property_read_u64(np, "32-bit-extended-addr",
+	                                &info->extended_addr);
+	        if (error) {
+	                dev_err(&pdev->dev,
+	                        "Miss 32-bit-extended-addr property\n");
+	                return error;
+	        }
+	}
+
 	ndev->features = info->net_features;
 	ndev->hw_features = info->net_hw_features;
 
-- 
2.25.1

